### PR TITLE
(Test-only) Add tests for new `ImageFilter.dilate`/`ImageFilter.erode` in flutter engine

### PR DESCRIPTION
**Pull request by fzyzcjy**
*Replace this paragraph with a description of what this PR is changing or adding, and why. Consider including before/after screenshots.*
Please see https://github.com/flutter/flutter/issues/100830

*List which issues are fixed by this PR. You must list at least one issue.*
https://github.com/flutter/flutter/issues/100830

Related: https://github.com/flutter/engine/pull/32334

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by fzyzcjy** [comment]
This should fail until https://github.com/flutter/engine/pull/32334 is merged and flutter/flutter repo uses the latest flutter/engine.

**Comment by fzyzcjy** [comment]
We know dilate/erode is not implemented for the web platform, just like the compose filter. So what should I do (skip tests when in web?)?

```
══╡ EXCEPTION CAUGHT BY FLUTTER TEST FRAMEWORK ╞════════════════════════════════════════════════════
The following UnimplementedError was thrown running a test:
ImageFilter.dilate not implemented for web platform.

When the exception was thrown, this was the stack:
../dart-sdk/lib/_internal/js_dev_runtime/private/ddc_runtime/errors.dart 251:49  throw_
../lib/ui/painting.dart 411:5                                                    dilate
image_filter_test.dart.js 428:133                                                <fn>
../dart-sdk/lib/_internal/js_dev_runtime/patch/async_patch.dart 84:54            runBody
../dart-sdk/lib/_internal/js_dev_runtime/patch/async_patch.dart 123:5            _async
image_filter_test.dart.js 427:80                                                 <fn>
../packages/flutter_test/src/matchers.dart.js 4522:19                            <fn>
```

**Comment by dnfield** [comment]
Skip these on web with a reference to the bug filed to implement them for web.

**Comment by fzyzcjy** [comment]
Thanks, I will do that.

**Comment by goderbauer** [comment]
@fzyzcjy Can you update the tests per the suggestion above to make the tests happy?

**Comment by fzyzcjy** [comment]
Sure. I forgot it

**Comment by goderbauer** [comment]
@fzyzcjy Can you take a look at the golden images linked in the previous comment to confirm that that's what you expected them to look like? One of the images looks empty, which seems odd...

**Comment by fzyzcjy** [comment]
@goderbauer I guess it is right, just b/c it erodes too much. I have updated the test so it will be less strange.

**Comment by fzyzcjy** [comment]
The gold looks better. (Notice it has very very thin lines)

File: packages/flutter/test/widgets/image_filter_test.dart [hunk 1]
```diff
@@ -29,6 +30,37 @@ void main() {
...
+    );
+  }, skip: kIsWeb); // https://github.com/flutter/flutter/issues/101874
+
+  testWidgets('Image filter - erode', (WidgetTester tester) async {
+    await tester.pumpWidget(
+      RepaintBoundary(
+        child: ImageFiltered(
+          // Do not erode too much, otherwise we will see nothing left.
+          imageFilter: ImageFilter.erode(radiusX: 1.0, radiusY: 1.0),
+          child: const Placeholder(),
```
Comment (flar) [inline packages/flutter/test/widgets/image_filter_test.dart:54]: The place holder by default draws a box with diagonals with a stroke width of 2 and doesn't survive the erode process. Try increasing the stroke width on both to a reasonable size so it doesn't erode completely for the golden image.

**Review by flar — CHANGES_REQUESTED** [review]

File: packages/flutter/test/widgets/image_filter_test.dart (hunk 1 above)
Comment (flar) [reply packages/flutter/test/widgets/image_filter_test.dart:54]: Actually, the dilate filter shows up fine in its golden, it's definitely more than 2 pixels wide, but the erode filter creates a blank image.

Comment (fzyzcjy) [reply packages/flutter/test/widgets/image_filter_test.dart]: stroke width modified.

**Comment by flar** [comment]
> Thanks, @fzyzcjy
> 
> @dnfield or @flar could you review this one (and also approve the goldens) since you reviewed the upstream PR in the engine?

I still don't see anything rendered in the erode test even if I click on it to see it full size. Am I clicking on the wrong link?

Actually, it looks like I was reviewing the goldens from before I requested a larger stroke width. Maybe they haven't been updated yet?

**Comment by fzyzcjy** [comment]
> I still don't see anything rendered in the erode test even if I click on it to see it full size. Am I clicking on the wrong link?
Actually, it looks like I was reviewing the goldens from before I requested a larger stroke width. Maybe they haven't been updated yet?

I guess the CI is even not executed after modifying strokewidth...

image

**Comment by fzyzcjy** [comment]
I guess the failure (and seems skipped executing all tests related to golden) is not caused by me. So maybe need to wait

**Comment by flar** [comment]
I would just LGTM to let it merge, but I'm worried it may merge with bad goldens and cause problems down the line.

**Comment by fzyzcjy** [comment]
Take your time. Since this is a test-only PR I am not that urgent :)

**Comment by flar** [comment]
It looks like you need to rebase in order to get the ci.yaml to gel with the latest CI recipes and then hopefully final goldens will be generated.

**Comment by fzyzcjy** [comment]
Now seems better

**Review by flar — APPROVED** [review]
LGTM

**Review by goderbauer — APPROVED** [review]
LGTM

**Comment by darrenaustin** [comment]
Where the goldens for this approved? It looks like this is breaking the build:

https://logs.chromium.org/logs/flutter/buildbucket/cr-buildbucket/8815144942697521329/+/u/run_test.dart_for_framework_tests_shard_and_subshard_widgets/test_stdout

```
The following _Exception was thrown while running async test code:
Exception: Skia Gold received an unapproved image in post-submit
testing. Golden file images in flutter/flutter are triaged
in pre-submit during code review for the given PR.

Visit https://flutter-gold.skia.org// to view and approve
the image(s), or revert the associated change. For more
information, visit the wiki:
https://github.com/flutter/flutter/wiki/Writing-a-golden-file-test-for-package:flutter

Debug information for Gold:
stdout: Given image with hash 94028cfacbe41a55af47edcddeff6540 for test widgets.image_filter_erode
Untriaged or negative image:
https://flutter-gold.skia.org/detail?test=widgets.image_filter_erode&digest=94028cfacbe41a55af47edcddeff6540


stderr: Test: widgets.image_filter_erode FAIL

When the exception was thrown, this was the stack:
#0      SkiaGoldClient.imgtestAdd (package:flutter_goldens_client/skia_client.dart:216:7)
<asynchronous suspension>
<asynchronous suspension>
(elided one frame from package:stack_trace)
```

Can we just approve the new goldens? If not I will need to revert this to get the tree green again.

**Comment by fzyzcjy** [comment]
I do not have permission to approve the goldens, but seems that @flar has said that the goldens are ok

**Comment by darrenaustin** [comment]
It looks like it has been resolved and the tree is green again, so it's all good now.

**Comment by flar** [comment]
I did approve the goldens, so I'm not sure what happened.

**Pull request merged**